Storefront - Api for fetching details of a store by its code

coolprobn opened the pull request:
- Add api for getting details of store
- Add relevant routes
- Add serializer for store

damianlegawiec commented on api/app/serializers/spree/v2/storefront/store_serializer.rb:
```diff
@@ -0,0 +1,13 @@
+
+        attributes :name, :url, :meta_description, :meta_keywords, :seo_title, :mail_from_address, :default_currency, :code, :default,
+                   :supported_currencies, :facebook, :twitter, :instagram, :default_locale, :customer_support_email, :default_country_id, :description,
+                   :address, :contact_phone, :contact_email, :new_order_notifications_email
```
@coolprobn I think we can skip `new_order_notifications_email` - that's usually an internal email address and should not be made public, as well as `mail_from_address` and `code`.

Besides that - great work! Thank you!

coolprobn replied:
Sure, I will make required changes and push again. Thank you for pointing them out.

coolprobn replied:
This has been resolved. Thank you!

coolprobn committed:
Storefront - Api for store

- Add api for getting details of store
- Add relevant routes
- Add serializer for store

damianlegawiec commented on api/app/controllers/spree/api/v2/storefront/stores_controller.rb:
```diff
@@ -0,0 +1,23 @@
+
+          private
+
+          def resource
+            Spree::Store.find_by(params[:id])
+          end
+
```
```suggestion
          def scope
            Spree::Store.accessible_by(current_ability, :show)
          end

          def resource
            scope.find_by(code: params[:id])
          end
```

Probably it's better to use code rather than expose the ID

coolprobn replied:
I agree. I have updated the code to reflect what you have suggested. Thanks!

coolprobn committed:
Storefront Api for store

- Change route to fetch store by code rather than id
- Add read ability for a store
- Update code in store controller to find store by code

damianlegawiec commented on api/config/routes.rb:
```diff
@@ -161,6 +161,7 @@
         get '/order_status/:number', to: 'order_status#show', as: :order_status
         resources :products, only: %i[index show]
         resources :taxons,   only: %i[index show], id: /.+/
+        get '/stores/:code', to: 'stores#show', as: :store
```
@coolprobn I think is somehow conflicting with API v1 routes - hence the failing tests, let's leave it as 

```suggestion
        resources :stores, only: :show
```

coolprobn replied:
Test seems to be failing after we added ability of the user to view store without login so they can access particular store data by it's code.
In V1 API, user needs to be authorized to view the store details, but for the storefront we need the data of store to be accessible without logging in.
Maybe we should update the tests, what would you suggest?

coolprobn replied:
Hello @damianlegawiec
Any suggestions?

damianlegawiec commented on api/app/controllers/spree/api/v2/storefront/stores_controller.rb:
```diff
@@ -0,0 +1,27 @@
+          private
+
+          def scope
+            Spree::Store.accessible_by(current_ability, :show)
```
```suggestion
            Spree::Store
```

damianlegawiec commented on core/app/models/spree/ability.rb:
```diff
@@ -59,6 +59,7 @@ def initialize(user)
         can :read, Taxonomy
         can :read, Variant
         can :read, Zone
+        can :read, Store
```
```suggestion
```

damianlegawiec reviewed (commented):
@coolprobn let's follow the approach that I've left in the comments - so expose this by default in API v2 and not change the logic of API v1

coolprobn committed:
Remove store show from ability

- Remove show action of store from ability
- Let store be fetched by everyone irrelevant of their ability

coolprobn commented:
@damianlegawiec I have made the changes as suggested. Thank you!

damianlegawiec reviewed (approved):
Thanks @coolprobn 👍
